Elastic config description (#1108)
ezsystems/developer-documentation: docs/guide/search/elastic.md

    ### Configure security

    Elasticsearch instances support `basic` and `api_key` authentication methods.
    You select authentication type and configure the settings under the `authentication` key. By default, authentication is disabled:

    ``` yaml
    <connection_name>:
        # ...
        authentication:
            type: null
    ```

    If you connect to Elasticsearch hosts outside of your local network, you might also need to configure SSL encryption.

    #### Configure basic authentication

    If your Elasticsearch server is protected by HTTP authentication, you must provide eZ Platform with the credentials.
    When using basic authentication, you must pass the following parameters:

    ``` yaml
    <connection_name>
        # ...
        authentication:
            type: basic
            credentials: ['<user_name', '<password>']
    ```

    For example:

    ``` yaml
    ezplatform_elastic_search_engine:
        connections:
        	cloud:
            	debug: true
            	elastic_cloud_id: 	'test:ZWFzdHVzMi5henVyZS5lbGFzdGljLWNsb3VkLmNvbTo5MjQzJGUwZ'
            	authentication:
                	type: basic

>                 	credentials: ['elastic', '1htFY83VvX2JRDw88MOkOejk']

Reviewer: This is not a valid YAML.

    ```

    For more information, see [Elasticsearch: SSL Encyption](https://www.elastic.co/guide/en/elasticsearch/client/php-api/7.x/security.html#_ssl_encryption_2).

    #### Enable debugging

    In a staging environment, you can log messages about the status of communication with Elasticsearch.
    You can then use Symfony Profiler to review the logs.

    By default, debugging is disabled. To enable debugging, you can toggle either of the following two settings:

    ``` yaml
    <connection_name>:
        # ...
        debug:	<true/false>
        trace:	<true/false>
    ```

    - `debug` logs basic information about a request, such as request status and time.
    - `trace` logs additional information, such as steps to reproduce an exact copy of a query.

    !!! tip

        Make sure that you disable debugging in a production environment.

    ### Configure field type mapping templates

    Before you can re-index the eZ Platform data, so that Elasticsearch can search through its contents, you must define an index template.
    Templates instruct Elasticsearch to recognize eZ Platform Fields as specific data types, based on, for example, a field name.

>     They help you prevent Elasticsearch from using the dynamic field mapping feature to create type mappings automatically.

Reviewer: IMHO this statement is redundant.

    You can create several field type mapping templates for each index, for example, to define settings that are specific for different languages.
    When you establish a relationship between a field mapping template and a connection, you can apply several templates, too.

    #### Define a template

    To define a field mapping template, you must provide a number of settings under the `index_templates` key.
    The structure of the template is as follows:

    ``` yaml
    ez_platform_elastic_search_engine:
        # ...
        index_templates:
            <index_template_name>:
                patterns:
                # ...
                settings:
                # ...
                mappings:
                # ...
    ```

    Set a unique name for the template and configure the following keys:

    - `patterns` - A list of wildcards that Elasticsearch uses to match the field mapping template to an index.
    If you use the `_location_*` wildcard as a pattern, the settings and mappings listed below apply

>     to all indexes whose names are prefixed with the word "location".

Reviewer: We are missing here documentation for index naming convention. User will not be able to to provide a valid index name pattern without it.

                        # ...
    ```

    You can also set a boosting factor for a specific field.
    Boosting increases the relevance of hits, for example making keywords from the title more relevant than the ones from other places of the document.
    Set the boosting factor under the `properties` key:

    ``` yaml
    ez_platform_elastic_search_engine:
        # ...
        index_templates:
            default:
                # ...
                mappings:
                    properties:
                        content_name_s:
                            boost: 2.0
                    # ...
    ```

    You can even copy contents of existing fields, process them and then paste into another field, which than can be queried:

    ``` yaml
    ez_platform_elastic_search_engine:
        # ...
        index_templates:
            default:
                # ...
                mappings:
                    properties:
                        user_first_name_s:
                            type: keyword
                            normalizer: lowercase_normalizer
                            copy_to: custom_field
                    # ...
    ```

>     ### Bind templates with connections

Reviewer: IMHO we should link here to ES documentation to explain how it behave in case of multiple templates: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates-v1.html#multiple-templates-v1

    Once you have created the field mapping template(s), you must establish a relationship between the templates and a connection. You do this by adding the "index_templates" key to a connection definition.

    If your configuration file contains several connection definitions, you can reuse the same template for different connections.
    If you have several index templates, you can apply different combinations of templates to different connections.

    ``` yaml
    <connection_name>:
        # ...
        index_templates:
            - default
            - default_en_us
    ```

    ## Step 5: Push the templates

>     Push the templates to the Elasticsearch engine by running the following command:

Reviewer: We should make it clear that this step should be done for each of the defined connection.

    ``` bash
    php bin/console ezplatform:elasticsearch:put-index-template
    ```

    You can modify the behavior of the command with a number of switches. Use the `-h` switch to display a complete list of available options.

    ## Step 6: Reindex the database

    After creating index templates, run the following command to reindex your data:

    ``` bash
    php bin/console ezplatform:reindex
    ```

    !!! caution "Risks of premature indexing"

>         Do not reindex your data before you create custom templates.

Reviewer: ```suggestion
        Do not reindex your data before you create index templates.
```

    They help you prevent Elasticsearch from using the dynamic field mapping feature to create type mappings automatically.
    You can create several field type mapping templates for each index, for example, to define settings that are specific for different languages.
    When you establish a relationship between a field mapping template and a connection, you can apply several templates, too.

    #### Defining a template

    To define a field mapping template, you must provide a number of settings under the `index_templates` key.
    The structure of the template is as follows:

    ``` yaml
    ez_platform_elastic_search_engine:
        # ...
        index_templates:
            <index_template_name>:
                patterns:
                # ...
                settings:
                # ...
                mappings:
                # ...
    ```

    Set a unique name for the template and configure the following keys:

    - `patterns` - A list of wildcards that Elasticsearch uses to match the field mapping template to an index.
    Index names use the following pattern:

        `<repository>_<document_type>_<language_code>_<content_type_id>`

>         By default, repository name is set to `default`, however, in the context of a site access,

Author: ```suggestion
        By default, repository name is set to `default`, however, in the context of an eZ Platform instance,
```

# Elasticsearch search engine

!!! enterprise

    Elasticsearch is an open-source, distributed, Java-based search engine that responds to queries
    in real-time and is easily scalable in reaction to changing processing needs.

    With this document you learn to configure Elastic and eZ Platform to work together.
    For a detailed description of advanced settings that you might require in a specific
    production environment, see the documentation provided by Elastic, beginning with the [Set up Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/setup.html) section.

    !!! tip "**Prerequisite**"

        Before you proceed to configuring the integration between Elasticsearch and eZ Platform, make sure
      	that you install a standalone instance of Elasticsearch locally and experiment with it
      	to become familiar with how indexing, filtering and queries work.

>     ## How to set up the Elasticsearch search engine

Reviewer 2: How about removing this heading and bumping all other headings one level up? That will give a nicer TOC

    Elasticsearch integration requires that you do the following steps:

    ### Step 1: Download and install Elasticsearch

    Download an Elasticsearch installation package for the system platform of your choice and install it on a dedicated machine using out-of-the-box settings.
    For example, download the `tar.gz` file and execute the `bin/elasticsearch` command.

    For more information, see [Install Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/install-elasticsearch.html).

    !!! note

>         eZ Platform is optimised to work with Elasticsearch in version 7.7.

Author: @adamwojs Would eZ work with ES in other versions? Could I say "supports ES 7.7 and up"?

    	    intranet:
    	      debug: true
    	      hosts:
    	        - "c.elasticsearch.loc:9200"

    	  default_connection: simple
    ```

    ###### Configure the multi-node cluster behaviour

    When you configure a cluster-based connection, and the cluster consists of many nodes, you can
    choose strategies that govern how the cluster reacts to changing operating conditions, or how
    workload is distributed among the nodes.

    Connection pool

    With this setting you decide how a list of hosts that form a cluster is managed.
    The list of active hosts tends to change in time, due to different reasons, such as connectivity
    issues, host malfunction, or the fact that you add new hosts to the cluster to increase
    its performance.
    By default, the `StaticNoPingConnectionPool` setting is used.

    You can change the default setting with the following key:

    ``` yaml
    <connection_name>:
      # ...
      connection_pool: Elasticsearch\ConnectionPool\<connection_pool_name>
    ```

    For more information and a list of available choices, see [Connection pool](https://www.elastic.co/guide/en/elasticsearch/client/php-api/7.x/connection_pool.html).

>     One of possible choices is the `sniffingConnectionPool` setting, which provides increased flexibility.

Reviewer 2: Add why you select this setting out of the others. Is it particularly recommended for Platform? If not, maybe not worth mentioning

    This connection pool setting requires that you provide a list of seed hosts.
    Elasticsearch then queries seed hosts for cluster topology and pulls a list of available hosts.
    However, this setting might result in a large amount of overhead.

    !!! tip "Load tests recommendation"

    		It is recommended that you to perform load tests to check whether the change does not negatively impact the performance of your environment.

>     Connection selector

Reviewer 2: Heading?

>     When the cluster consists of many hosts, with this setting you decide what strategy is used to pick a node to send a query requests to.

Reviewer 2: ```suggestion
    When the cluster consists of many hosts, the `connection_selector` setting decides what strategy is used to pick a node to send a query requests to.
```

    By default, the `RoundRobinSelector` setting is used.

>     If you prefer a different strategy, or have created your own, proprietary strategy, you can change the default setting with the following key:

Reviewer 2: ```suggestion
    If you prefer a different strategy, or have created your own, custom strategy, you can change the default setting with the following key:
```

    ``` yaml
    <connection_name>:
      # ...
      connection_selector: Elasticsearch\ConnectionPool\Selectors\<selector_name>
    ```

    For more information and a list of available choices, see [Selectors](https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/selectors.html).

>     Number of retries

Reviewer 2: Heading

>     With this setting you can configure the number of attempts that eZ platform makes to connect

Reviewer 2: ```suggestion
    The `retries` setting configures the number of attempts that eZ Platform makes to connect
```

    to the nodes of the cluster before it throws an exception.

>     By default, the `null` setting is used, which means that the number of retries equals to the number of nodes in the cluster.

Reviewer 2: ```suggestion
    By default, `null` is used, which means that the number of retries equals to the number of nodes in the cluster.
```

    ``` yaml
    <connection_name>:
      # ...
      retries: null
    ```

    Depending on the connection pool that you select, eZ Platform's reaction to reaching the maximum number of retries might differ.

    For more information, see [Set retries](https://www.elastic.co/guide/en/elasticsearch/client/php-api/7.x/configuration.html#_set_retries).

    ##### Configure Elasticsearch Cloud

>     As an alternative, you might decide to use Elasticsearch Cloud, a commercial SaaS solution.

Reviewer 2: ```suggestion
    As an alternative to using your own cluster, you can use Elasticsearch Cloud, a commercial SaaS solution.
```
?

    With Elasticsearch Cloud you do not have to build or manage your own Elasticsearch cluster.
    Also, you do all the configuration and administration in a graphical user interface.

    To connect to a cloud solution with eZ Platform, you must set the `elastic_cloud_id` parameter by
    providing an alphanumerical ID string that you obtain from the cloud's user interface, for example:

    ``` yaml
    <connection_name>:
      elastic_cloud_id: 'production:ZWFzdHVzMi5henVyZS5lbGFzdGljLWNsb3VkLmNvbTo5MjQzJGUwZ'
    ```

    With the ID set, you must configure authentication to be able to access the remote environment.

    ##### Configure security

    Elasticsearch instances support `basic` and `api_key` authentication methods.
    You select authentication type and configure the settings under the `authentication` key. By default, authentication is disabled:

    ``` yaml
    <connection_name>:
      # ...
    	authentication:
    	  type: null
    ```

    If you connect to Elasticsearch hosts outside of your local network, you might also need to configure SSL encryption.

>     #### Configure basic authentication

Reviewer 2: Maybe a #### level Authentication heading, with basic and API below it?

>     If your Elasticsearch server is protected by HTTP authentication, you must provide eZ Platform with the credentials, so that eZ Platform requests can be authenticated by the server.

Reviewer 2: ```suggestion
    If your Elasticsearch server is protected by HTTP authentication, you must provide eZ Platform with the credentials.
```

>     Basic authentication support requires that you pass a set of parameters that is similar to the following example:

Reviewer 2: ```suggestion
    When using basic authentication you must pass the following parameters:
```

    ``` yaml
    <connection_name>
      # ...
    	  authentication:
    		  type: basic
    		  credentials: ['<user_name', '<password>']
    ```

    For example:

    ``` yaml
    ezplatform_elastic_search_engine:
      connections:
        cloud:
          debug: true
          elastic_cloud_id: 	'test:ZWFzdHVzMi5henVyZS5lbGFzdGljLWNsb3VkLmNvbTo5MjQzJGUwZ'
          authentication:
    	      type: basic
    	      credentials: ['elastic', '1htFY83VvX2JRDw88MOkOejk']
    ```

    #### Configure API key authentication

>     If your Elasticsearch cluster is protected by API keys, you must use the values described below to

Reviewer 2: ```suggestion
    If your Elasticsearch cluster is protected by API keys, you must provide the key and secret in authentication configuration to
```
Something like this - IMHO there's too much of a jump between "described below" and the actual example.

    connect eZ Platform with the cluster. With API key authentication you can define different
    authorisation levels, such as [`create_index`, `index`, etc.](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/security-privileges.html#privileges-list-indices). Such approach
    proves useful if the cluster is available to the public.

    For more information, see [Create API key](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/security-api-create-api-key.html).

>     API key authentication support requires that you provide the following set of parameters to

Reviewer 2: ```suggestion
    When using API key authentication you must pass the following parameters to
```

    authenticate access to the cluster:

    ``` yaml
    <connection_name>:
      # ...
      authentication:
        type: api_key
        credentials: ['<api_key>', '<api_secret>']
    ```

    For example:

    ``` yaml
    ezplatform_elastic_search_engine:
      connections:
        cloud:
          debug: true
          elastic_cloud_id: 	'test:ZWFzdHVzMi5henVyZS5lbGFzdGljLWNsb3VkLmNvbTo5MjQzJGUwZ'
          authentication:
            type: api_key
            credentials: ['8Ek5f3IBGQlWj6v4M7zG', 'rmI6IechSnSJymWJ4LZqUw']
    ```

    #### Configure SSL

    When you need to protect your communication with the Elasticsearch server, you can use SSL encryption.
    When configuring SSL for your internal infrastructure, you can use your own certificates.

>     You configure ssl by passing the path-passwords pairs for both the certificate and the certificate key.

Reviewer 2: ```suggestion
    Configure SSL by passing the path-passwords pairs for both the certificate and the certificate key.
```

    For example:

    ``` yaml
    ezplatform_elastic_search_engine:
      connections:
        cloud_with_ssl:
          debug: true
          elastic_cloud_id: 	'test:ZWFzdHVzMi5henVyZS5lbGFzdGljLWNsb3VkLmNvbTo5MjQzJGUwZ'
          authentication:
            type: api_key
            credentials: ['8Ek5f3IBGQlWj6v4M7zG', 'rmI6IechSnSJymWJ4LZqUw']
          ssl:
            cert:
              path: '/path/to/cert.pem'
              pass: ~
            cert_key:
              path: '/path/to/cert-key'
              pass: ~
    ```

>     After you have configured SLL, you can still disable it, for example, when the certificates expire, or you are migrating to a new set of certificates.

Reviewer 2: ```suggestion
    After you have configured SLL, you can still disable it, for example when the certificates expire, or you are migrating to a new set of certificates.
```

>     To do this, you must pass the following setting under the `ssl` key:

Reviewer 2: ```suggestion
    To do this, pass the following setting under the `ssl` key:
```

    ``` yaml
    verification: false
    ```

    For more information, see [Elasticsearch: SSL Encyption](https://www.elastic.co/guide/en/elasticsearch/client/php-api/7.x/security.html#_ssl_encryption_2).

    #### Enable debugging

>     In a staging environment, you can decide that eZ Platform populates its logs with messages about the status of communication with Elasticsearch. You can then use Symfony Profiler to review the logs.

Reviewer 2: ```suggestion
    In a staging environment, you can log messages about the status of communication with Elasticsearch. You can then use Symfony Profiler to review the logs.
```

    By default, debugging is disabled. To enable debugging, you can toggle either of the following two settings:

    ``` yaml
    <connection_name>:
      # ...
      debug:	<true/false>
      trace:	<true/false>
    ```

>     Enabling the `debug` setting results in logging basic information about a request, such as request status and time.
>     Enabling the `trace` setting results in logging additional information, such as steps to reproduce an an exact copy of a query.

Reviewer 2: ```suggestion
    - `debug` logs basic information about a request, such as request status and time.
    - `trace` logs additional information, such as steps to reproduce an an exact copy of a query.
```
?

    !!! tip

    	Make sure that you disable debugging in a production environment.

>     #### Configure field type mappings

Reviewer 2: If this refers to Field Types as a Platform domain term, they should be capitalized (throughout the whole PR)

    Before you can re-index the eZ Platform data, so that Elasticsearch can search through its contents, you must define an index template.

>     Templates instruct Elasticsearch to recognise eZ Platform fields as specific data types, based on, for example, a field name.

Reviewer 2: ```suggestion
    Templates instruct Elasticsearch to recognise eZ Platform Fields as specific data types, based on, for example, a field name.
```

    They help you prevent Elasticsearch from using the dynamic field mapping feature to create type mappings automatically.

>     You can create several field type mapping templates for each index, for example, to define settings that are specific for different languages.

Reviewer 2: ```suggestion
    You can create several field type mapping templates for each index, for example to define settings that are specific for different languages.
```

    When you establish a relationship between a field mapping template and a connection, you can apply several templates, too.

    ##### Define a template

    To define a field mapping template, you must provide a number of settings under the `index_templates` key.
    The structure of the template is as follows:

    ``` yaml
    ez_platform_elastic_search_engine:
      # ...
      index_templates:
        <index_template_name>:
            patterns:
              # ...
            settings:
              # ...
            mappings:            
    ```

>     You first set a unique name for the template, then the keys, which have the following meaning:

Reviewer 2: ```suggestion
    Set a unique name for the template and configure the following keys:
```

    - `patterns` - A list of wildcards that Elasticsearch uses to match the field mapping template to an index.

>     If you use the `_location_*` wildcard as a pattern, the settings and mappings listed belowapply to all indexes, whose names are prefixed with the word "location".

Reviewer 2: ```suggestion
    If you use the `_location_*` wildcard as a pattern, the settings and mappings listed below apply to all indexes whose names are prefixed with the word "location".
```

>       The `patterns` setting proves useful when your data contains content in different languages, you can create index templates with settings that apply to a specific language only, for example, to eliminate stop words from the index, or help divide concatenations.

Reviewer 2: ```suggestion
      You can use the `patterns` setting when your data contains content in different languages.
      You can create index templates with settings that apply to a specific language only,
      for example to eliminate stop words from the index, or help divide concatenations.
```

    ##### Fine-tune the search results

    Your search results can be adjusted by configuring additional parameters. For a list of available mapping parameters and their usage, see [Elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/mapping-params.html).

    For example, you can apply a mapping parameter, in this case, a normalizer, to a specific mapping under the `dynamic_templates` key:

    ``` yaml
    ez_platform_elastic_search_engine:
      # ...
      index_templates:
        default:
          # ...
          mappings:
            # ...
            dynamic_templates:
              - ez_string:
                  match: "*_s"
                  mapping:
                    type: keyword
                    normalizer: lowercase_normalizer
              # ...
    ```

    You can also set a boosting factor for a specific field.

>     Boosting increases the relevance of hits, for example, to help you account for the fact that keywords from the title are more relevant than the ones from other places of the document.

Reviewer 2: ```suggestion
    Boosting increases the relevance of hits, for example making keywords from the title more relevant than the ones from other places of the document.
```

>     You set the boosting factor under the `properties` key:

Reviewer 2: ```suggestion
    Set the boosting factor under the `properties` key:
```

                boost: 2.0
              # ...
    ```

    You can even copy contents of existing fields, process them and then paste into another field, which than can be queried:

    ``` yaml
    ez_platform_elastic_search_engine:
      # ...
      index_templates:
        default:
          # ...
          mappings:
            properties:
              user_first_name_s:
                type: keyword
                normalizer: lowercase_normalizer
                copy_to: custom_field
              # ...
    ```

    ##### Bind templates with connections

    Once you have created the field mapping template(s), you must establish a relationship between the templates and a connection. You do this by adding a "index_templates" key to a connection definition.

    If your configuration file contains several connection definitions, you can reuse the same template for different connections.
    If you have several index templates, you can apply different combinations of templates to different connections.

    ``` yaml
    <connection_name>:
      # ...
      index_templates:
        - default
        - default_en_us
    ```

    ### Step 5: Push the templates

>     You push the templates to the Elasticsearch engine by running the following command:

Reviewer 2: ```suggestion
    Push the templates to the Elasticsearch engine by running the following command:
```

    ``` bash

>     run php bin/console ezplatform:elasticsearch:put-index-template

Reviewer 2: ```suggestion
    php bin/console ezplatform:elasticsearch:put-index-template
```
I think?

    ```

>     You can modify the behaviour of the command with a number of switches:

Reviewer 2: If this is the output of `--help`, we don't need to repeat this info here

    | Switch | meaning |
    |---|---|
    | -c, --connection[=CONNECTION] | Elasticsearch connection name |
    | --overwrite | Overwrite existing index template |
    | -h, --help | Display a help message |
    | -q, --quiet | Do not output messages |
    | -V, --version | Display application version |
    | --ansi | Force ANSI output |
    | --no-ansi | Disable ANSI output |
    | -n, --no-interaction | Do not ask interactive questions |
    | -e, --env=ENV | Environment name (default value is "dev") |
    | --no-debug | Switch off debug mode |
    | --siteaccess[=SITEACCESS] | SiteAccess to be uses for operations. If not provided, default siteaccess is used |
    | -v/vv/vvv, --verbose | Increase the verbosity of messages (1 is normal output, 2 is verbose output, 3 is debug) |
    |---|---|

    ### Step 6: Reindex the database

>     Once you have created index templates that satisfy your needs, run the following command to reindex your data:

Reviewer 2: ```suggestion
    After creating index templates, run the following command to reindex your data:
```

    ``` bash

>     run php bin/console ezplatform:reindex

Reviewer 2: ```suggestion
    php bin/console ezplatform:reindex
```
I think?
